src/drt/src/TritonRoute.cpp

```diff
@@ -986,6 +964,110 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
   reportDRC(filename, markers, requiredDrcBox);
 }
 
+void TritonRoute::processBTermsAboveTopLayer()
+{
+  odb::dbTech* tech = db_->getTech();
+  odb::dbBlock* block = db_->getChip()->getBlock();
+
+  odb::dbTechLayer* top_tech_layer
+      = tech->findLayer(TOP_ROUTING_LAYER_NAME.c_str());
+  if (top_tech_layer != nullptr) {
+    int top_layer_idx = top_tech_layer->getRoutingLevel();
+    for (auto bterm : block->getBTerms()) {
+      std::vector<odb::dbTechLayer*> bterm_layers;
+      for (auto bpin : bterm->getBPins()) {
+        for (auto box : bpin->getBoxes()) {
+          bterm_layers.push_back(box->getTechLayer());
+        }
+      }
+      int bterm_bottom_layer_idx = bterm_layers[0]->getRoutingLevel();
+      for (auto layer : bterm_layers) {
+        bterm_bottom_layer_idx
+            = std::min(bterm_bottom_layer_idx, layer->getRoutingLevel());
+      }
+
+      if (bterm_bottom_layer_idx > top_layer_idx) {
+        stackVias(bterm, top_layer_idx, bterm_bottom_layer_idx);
+      }
+    }
+  }
+}
+
+void TritonRoute::stackVias(odb::dbBTerm* bterm,
+                            int top_layer_idx,
+                            int bterm_bottom_layer_idx)
+{
+  odb::dbTech* tech = db_->getTech();
+  auto fr_tech = getDesign()->getTech();
+  odb::dbSet<odb::dbTechVia> vias = tech->getVias();
+  std::map<int, odb::dbTechVia*> default_vias;
+
+  for (auto layer : tech->getLayers()) {
+    if (layer->getType() == odb::dbTechLayerType::CUT) {
+      frLayer* fr_layer = fr_tech->getLayer(layer->getName());
+      frViaDef* via_def = fr_layer->getDefaultViaDef();
+      odb::dbTechVia* tech_via = tech->findVia(via_def->getName().c_str());
+      int via_bottom_layer_idx = tech_via->getBottomLayer()->getRoutingLevel();
+      default_vias[via_bottom_layer_idx] = tech_via;
+    }
+  }
+
+  if (default_vias.empty()) {
+    for (int i = top_layer_idx; i <= bterm_bottom_layer_idx; i++) {
+      for (odb::dbTechVia* via : vias) {
+        if (via->getBottomLayer()->getRoutingLevel() == i) {
+          default_vias[i] = via;
+          break;
+        }
+      }
+    }
+  }
+
+  // get bterm access points
+  std::vector<odb::dbAccessPoint*> access_points;
+  odb::Rect pin_rect;
+  for (odb::dbBPin* bpin : bterm->getBPins()) {
+    const std::vector<odb::dbAccessPoint*>& bpin_pas = bpin->getAccessPoints();
+    access_points.insert(access_points.end(), bpin_pas.begin(), bpin_pas.end());
+    pin_rect = bpin->getBBox();
+  }
```

Comment:
I might be wrong, so could you please modify the test case to have pin access?

Reply:
You're right; even when running pin_access before detailed_route, this function uses the pin center to place the vias. It's ok to use the center of the pin, since the use case of this new functionality is for the bumps in the upper layers, which usually have a large area. We can drop this part of the code and always use the pin center. @maliberty, do you agree with this approach?

Reply:
works for me